Redis sentinel support

czarneckid opened the pull request:
We're running some software behind Redis sentinel and needed to support sentinel in limits (via flask-limiter). I also added the proper bits for Travis CI to run correctly. 

There's another change I'd need to make flask-limiter to support directly setting a Storage backend since Redis sentinel is configured differently than a simple string, but that'll be in another PR to that project.

alisaifee commented:
@czarneckid. Thanks for bringing up this requirement and taking a shot at an implementation. 

The pull request will require some more work before it can be merged in - specifically in the areas of:
 - reducing code duplication between `RedisStorage` & `RedisSentinelStorage`
 - modifying the constructor of `RedisSentinalStorage` so that its still possible to construct the storage entirely from configuration values.

If you look at the example of `MemcachedStorage` it actually makes use of the `kwargs` `options` in case extra configuration is required from the lower level memcached client. Maybe something along those lines could work to instantiate the `Sentinal` instance. (I just realised that this isn't documented at all in the `Flask-Limiter` documentation - which is something I'll correct asap).

Another note: if you implement your own subclass of `limits.storage.Storage` in your own application code - it will still be picked up by `Flask-Limiter`. The following example shows how you can implement your own storage and upon initialization of the extension - it will be picked up based on the `storage_schema` specified by the configuration:

```
from limits.storage import MemoryStorage
from flask_limiter import Limiter
from flask import Flask

class MyStorage(MemoryStorage):
    STORAGE_SCHEME="mystorage"
    def __init__(self, uri, **kwargs):
        super(MyStorage, self).__init__(uri)

app = Flask(__name__)
limiter = Limiter(app, storage_uri="mystorage://") # this could be in the apps config file instead.

@limiter.limit("1/minute")
@app.route("/")
def index():
    return "42"

app.run()
```

czarneckid commented:
@alisaifee I think I've addressed your concerns with the last 2 commits. Let me know if you'd like to see other changes. I can certain squash these into a single commit.

alisaifee commented on limits/storage.py:
```diff
@@ -20,6 +20,8 @@
 from .errors import ConfigurationError
 from .util import get_dependency
 
+from redis.sentinel import Sentinel
```
Please remove as this could result in a failed import for someone using `memcache` or `in-memory` storage schemes.

alisaifee commented on limits/storage.py:
```diff
@@ -287,47 +361,100 @@ def acquire_entry(self, key, limit, expiry, no_add=False):
+            host, port = loc.split(":")
+            self.sentinel_configuration.append((host, int(port)))
+
+        self.sentinel = Sentinel(self.sentinel_configuration, socket_timeout=options.get('socket_timeout', 0.1))
```
Please use:
```
self.sentinel = get_dependency("redis.sentinel").Sentinel(
            self.sentinel_configuration,
            socket_timeout=options.get('socket_timeout', 0.1)
        )
```

alisaifee commented on limits/storage.py:
```diff
@@ -287,47 +361,100 @@ def acquire_entry(self, key, limit, expiry, no_add=False):
+    rate limit storage with redis sentinel as backend
+    """
+
+    STORAGE_SCHEME = "sentinel"
```
Could we use `redis+sentinel` as the scheme to be more explicit?

alisaifee commented on tests/test_storage.py:
```diff
@@ -24,6 +25,7 @@ def test_storage_string(self):
+        self.assertTrue(isinstance(storage_from_string("sentinel://localhost:26379", service_name='localhost-redis-sentinel'), RedisSentinelStorage))
         self.assertRaises(ConfigurationError, storage_from_string, "blah://")
 
     def test_storage_check(self):
```
Please add an assertion for the `check` method of `RedisSentinelStorage`

alisaifee commented on tests/test_storage.py:
```diff
@@ -7,11 +7,12 @@
+    MemoryStorage, RedisStorage, MemcachedStorage, RedisSentinelStorage,
     Storage, storage_from_string
 )
 
```
Please add a `flushall` call in the `setUp` method for storage tests to help ensure a clean state for the test runs.

alisaifee commented on requirements/redis-master.conf:
```diff
@@ -0,0 +1,938 @@
+# Redis configuration file example
```
Please move the sentinel `*.conf` files into a separate directory (maybe under `tests/redis-confs`)

alisaifee commented:
@czarneckid thanks for the quick turnaround. I've added a few minor comments - once you've addressed those I should be able to merge this PR into master (no need to squash the commits).

I will probably do some aesthetic refactoring once that is done and before I make a release to pypi. I'll make sure I maintain the interface as per the last few comments in this PR, however - so you should be able to start using your implementation.

czarneckid commented:
OK, all comments should be addressed now.

czarneckid commented:
The one other change I made in the code was the use of " over '. I saw that I hadn't followed that convention so I changed it.

alisaifee commented:
Thanks @czarneckid. You should start be able to use this implementation from `master` now. I'll update this issue when I've released this version to pypi.

alisaifee commented:
@czarneckid I'm seeing sporadic errors with lock acquisition when using redis with sentinel (shows up every now and then when you run [this](https://github.com/alisaifee/limits/blob/master/tests/test_storage.py#L183) test. Have you faced this problem? If you're using the moving window strategy you might even face it in production if you have high enough concurrency per route.

czarneckid commented:
I'll see if I can isolate that down to see what the error might be.

alisaifee commented:
Doesn't seem to be anything wrong with the `limits` or `redis-py` code - it's just that when acquiring the lock I've set a max `blocking_timeout` of 1 second (which in retrospect seems ridiculously high already). 

The test I mentioned spins 100 threads that are continuously accessing random routes from a pool of 100. That appears to create a situation where some requests are actually queued up long enough to hit the timeout.
